module/zfs/vdev_raidz.c

```diff
@@ -2774,6 +2854,238 @@ vdev_raidz_worst_error(raidz_row_t *rr)
+	for (int c = 0; c < samples; c++) {
+		vdev_t *cvd = vd->vdev_child[c];
+		if (cvd->vdev_prev_histo == NULL) {
+			mutex_enter(&cvd->vdev_stat_lock);
+			size_t size =
+			    sizeof (cvd->vdev_stat_ex.vsx_disk_histo[0]);
+			cvd->vdev_prev_histo = kmem_zalloc(size, KM_SLEEP);
+			memcpy(cvd->vdev_prev_histo,
+			    cvd->vdev_stat_ex.vsx_disk_histo[ZIO_TYPE_READ],
+			    size);
+			mutex_exit(&cvd->vdev_stat_lock);
+		}
+	}
+	uint64_t max = 0;
+	vdev_t *svd = NULL; /* suspect vdev */
+	uint_t sitouts = 0;
+	boolean_t skip = B_FALSE, svd_sitting = B_FALSE;
+	for (int c = 0; c < samples; c++) {
+		vdev_t *cvd = vd->vdev_child[c];
+		boolean_t sitting = vdev_sit_out_reads(cvd, 0) ||
+		    cvd->vdev_state != VDEV_STATE_HEALTHY;
+
+		if (sitting && ++sitouts >= vdev_get_nparity(vd)) {
+			/*
+			 * We can't sit out more disks than we have
+			 * parity
+			 */
+			skip = B_TRUE;
+		}
+		mutex_enter(&cvd->vdev_stat_lock);
+
+		uint64_t *prev_histo = cvd->vdev_prev_histo;
+		uint64_t *histo =
+		    cvd->vdev_stat_ex.vsx_disk_histo[ZIO_TYPE_READ];
+		if (skip) {
+			size_t size =
+			    sizeof (cvd->vdev_stat_ex.vsx_disk_histo[0]);
+			memcpy(prev_histo, histo, size);
+			mutex_exit(&cvd->vdev_stat_lock);
+			continue;
+		}
+		uint64_t count = 0;
+		lat_data[c] = 0;
+		for (int i = 0; i < VDEV_L_HISTO_BUCKETS; i++) {
+			uint64_t this_count = histo[i] - prev_histo[i];
+			lat_data[c] += (1ULL << i) * this_count;
+			count += this_count;
+		}
+		size_t size = sizeof (cvd->vdev_stat_ex.vsx_disk_histo[0]);
+		memcpy(prev_histo, histo, size);
+		mutex_exit(&cvd->vdev_stat_lock);
+		lat_data[c] /= MAX(1, count);
+
+		/* wait until all disks have been read from */
+		if (lat_data[c] == 0 && !sitting) {
+			skip = B_TRUE;
+			continue;
+		}
+
+		/* keep track of the vdev with largest value */
+		if (lat_data[c] > max) {
+			max = lat_data[c];
+			svd = cvd;
+			svd_sitting = sitting;
+		}
+	}
+
+	if (skip)
+		goto out;
+
+	qsort((void *)lat_data, samples, sizeof (uint64_t), latency_compare);
+	uint64_t iqr = 0;
+	uint64_t fence = latency_quartiles_fence(lat_data, samples, &iqr);
+
+	ASSERT3U(lat_data[samples - 1], ==, max);
+	if (max > fence && !svd_sitting) {
+		uint64_t incr = MAX(1, (max - fence) / iqr);
+		vd->vdev_outlier_count += incr;
+		if (vd->vdev_outlier_count >= samples) {
+			for (int c = 0; c < samples; c++) {
+				vdev_t *cvd = vd->vdev_child[c];
+				cvd->vdev_outlier_count -= 2;
+				cvd->vdev_outlier_count = MAX(0,
+				    cvd->vdev_outlier_count);
+			}
+			vd->vdev_outlier_count = 0;
+		}
+		/*
+		 * Keep track of how many times this child has had
+		 * an outlier read. A disk that persitently has a
+		 * higher than peers outlier count will be considered
+		 * a slow disk.
+		 */
+		svd->vdev_outlier_count += incr;
+		if (svd->vdev_outlier_count > LAT_OUTLIER_LIMIT) {
+			ASSERT0(svd->vdev_read_sit_out_expire);
+			vdev_raidz_sit_child(svd);
+			(void) zfs_ereport_post(FM_EREPORT_ZFS_DELAY,
+			    zio->io_spa, svd, NULL, NULL, 0);
```

Comment:
@tonyhutter mentioned that it would be useful to have some way to distinguish this delay event from the one posted in `zio_done()`.  We could infer it at the moment since this event doesn't include either a bookmark or a zio when posting, but something more explicit would be better.  Alternately, we could add a new FM_EREPORT_ZFS_SITOUT event.  That would have the advantage that the ZED won't take action on it.